Add new module for OrchardCore.KeyVault.Azure

JoshLefebvre opened the pull request:
Based on conversations had in https://github.com/OrchardCMS/OrchardCore/issues/6359

JoshLefebvre commented:
If this looks good, I would be happy to add some further documentation on how to use

deanmarcussen commented:
Looks great, will take a closer look tomorrow

+ 1 for some docs  :)

jtkech commented on src/OrchardCore/OrchardCore.KeyVault.Azure/Extensions/AzureKeyVaultWebHostBuilderExtension.cs:
```diff
@@ -0,0 +1,40 @@
+            return builder;
+        }
+
+
```
remove 2 extra lines

jtkech commented on src/OrchardCore/OrchardCore.KeyVault.Azure/OrchardCore.KeyVault.Azure.csproj:
```diff
@@ -0,0 +1,20 @@
+    <ProjectReference Include="..\OrchardCore.Abstractions\OrchardCore.Abstractions.csproj" />
+  </ItemGroup>
+
+
```
remove one extra line

jtkech commented on src/OrchardCore/OrchardCore.KeyVault.Azure/Services/CustomKeyVaultSecretManager.cs:
```diff
@@ -0,0 +1,23 @@
+            var key = secret.SecretIdentifier.Name.Replace("---", "_").Replace("--", ":");
+            return key;        
+        }
+
```
remove extra line

jtkech commented on src/OrchardCore.Build/Dependencies.props:
```diff
@@ -22,6 +22,7 @@
     <PackageManagement Include="MailKit" Version="2.6.0" />
     <PackageManagement Include="Markdig" Version="0.20.0" />
     <PackageManagement Include="MessagePack" Version="2.1.115" />
+    <PackageManagement Include="Microsoft.Extensions.Configuration.AzureKeyVault" Version="3.1.3" />
```
Move it to the `Dependencies.AspNetCore.props` file

    <PackageManagement Include="µ.Extensions.Configuration.AzureKeyVault" Version="$(AspNetCoreVersion)" />

For `AspNetCoreVersion` we currently use `3.1.5`, so we will have to check that it still builds.

jtkech commented on src/OrchardCore/OrchardCore.KeyVault.Azure/Services/CustomKeyVaultSecretManager.cs:
```diff
@@ -0,0 +1,23 @@
+        public override string GetKey(SecretBundle secret)
+        {
+            var key = secret.SecretIdentifier.Name.Replace("---", "_").Replace("--", ":");
+            return key;        
```
Maybe just return without using the key variable, but as you want as it may be useful for debugging ;)

jtkech commented on src/OrchardCore/OrchardCore.KeyVault.Azure/Services/CustomKeyVaultSecretManager.cs:
```diff
@@ -0,0 +1,23 @@
+    /// Examples:
+    /// Key Vault Input: "OrchardCore--OrchardCore---Shells---Database--ConnectionString".
+    /// Output: "OrchardCore:OrchardCore_Shells_Database:ConnectionString".
+    /// See https://github.com/OrchardCMS/OrchardCore/issues/6359.
```
Maybe this would have a better place in a `README.md` file under our `docs` folder

But that should not block this PR ;)

jtkech commented on src/OrchardCore/OrchardCore.KeyVault.Azure/Services/CustomKeyVaultSecretManager.cs:
```diff
@@ -0,0 +1,23 @@
+    /// Output: "OrchardCore:OrchardCore_Shells_Database:ConnectionString".
+    /// See https://github.com/OrchardCMS/OrchardCore/issues/6359.
+    /// </summary>
+    public class CustomKeyVaultSecretManager : DefaultKeyVaultSecretManager
```
Why not just naming it `AzureKeyVaultSecretManager`

JoshLefebvre commented:
Thanks @jtkech. Made all above requested changes and added the documentation.

jtkech commented:
@JoshLefebvre perfect i approved it

Just one last thing, maybe not the right place for the readme as `OrchardCore.KeyVault.Azure` is not a module but a core library. It does not bother me but i don't often work on documentation, i'm too lazy ;) so we need a confirmation, @agriffard can you confirm?

deanmarcussen commented:
Looks good @JoshLefebvre 

I fixed a couple of typos in the docs.

Can you move it to the core section as @jtkech suggested, and also it needs an entry in the `mkdocs.yml` or it won't display on readthedocs

xperiandri commented:
Is it ready to merge?

agriffard commented on src/docs/reference/core/KeyVault.Azure/README.md:
```diff
@@ -0,0 +1,40 @@
+# Azure Key Vault (`OrchardCore.Azure.KeyVault`)
+The Azure Key Vault configuration provider adds app configuration values from the Azure Key Vault in order to safeguard your cryptographic keys and secrets used by your app. It also contains custom override of the DefaultKeyVaultManger class that retrieves secrets from Azure Key Vault and translates ---
```
DefaultKeyVaultManger => DefaultKeyVaultManager

agriffard replied:
newline after title.

agriffard commented on src/docs/reference/core/KeyVault.Azure/README.md:
```diff
@@ -0,0 +1,40 @@
+See https://github.com/OrchardCMS/OrchardCore/issues/6359.
+
+## Configuration
+You'll need to specify the name of your Azure Key Vault and [register a service principle](https://docs.microsoft.com/en-us/azure/key-vault/general/group-permissions-for-apps) in Active Directory for accessing your key vault using an access control policy.
```
principle => principal

agriffard replied:
You'll need => You will need

agriffard commented on src/docs/reference/core/KeyVault.Azure/README.md:
````diff
@@ -0,0 +1,40 @@
+```
+
+!!! note
+    You should **never check in your client secret into source control** as this defeats the purpose of using a key vault in the first place. Instead set your client secret as an environment variable on your machine, create a seperate azurekeyvault.json file and add it to your `.gitignore`, or use user secrets.
````
Instead set => Instead, set

agriffard replied:
seperate => separate

sebastienros commented on src/OrchardCore/OrchardCore.KeyVault.Azure/Extensions/AzureKeyVaultWebHostBuilderExtension.cs:
```diff
@@ -0,0 +1,38 @@
+        /// </summary>
+        /// <param name="builder">The web host builder to configure.</param>
+        /// <returns>The web host builder.</returns>
+        public static IHostBuilder UseOrchardCoreAzureKeyVault(this IHostBuilder builder)
```
`AddOrchardCoreAzureKeyVault`

sebastienros commented on src/docs/reference/core/KeyVault.Azure/README.md:
```diff
@@ -0,0 +1,40 @@
+
+    public static IHost BuildHost(string[] args) =>
+        Host.CreateDefaultBuilder(args)
+            .UseOrchardCoreAzureKeyVault()
```
`AddOrchardCoreAzureKeyVault`

sebastienros commented on src/docs/reference/core/KeyVault.Azure/README.md:
```diff
@@ -0,0 +1,40 @@
+Output: "OrchardCore:OrchardCore_Shells_Database:ConnectionString".
+See https://github.com/OrchardCMS/OrchardCore/issues/6359.
+
+## Configuration
```
Can we have a small example that uses directly `AzureKeyVaultSecretManager` such that standard ASP.NET tutorials can be followrd?

JoshLefebvre replied:
Do you mean an example of IConfiguration? I have been using it to retrieve my media and database connection strings. For example in DatabaseShellConfigurationSources.cs:
![image](https://user-images.githubusercontent.com/22512098/86422868-51250980-bcac-11ea-83ab-684cf8bf0102.png)

Using the following Keys:
OrchardCore--OrchardCore---Shells---Database--ConnectionString
OrchardCore--OrchardCore---Shells---Database--DatabaseProvider (Not actually a secret, but required to get the section)

deanmarcussen replied:
What Seb was asking for is something that matches the getting started guides from the ASP.NET docs here https://docs.microsoft.com/en-us/aspnet/core/security/key-vault-configuration?view=aspnetcore-3.1

Where they are using the `AzureServiceTokenProvider` callbacks and such.

Which makes me also wonder if we should not be using the callback technique, rather than having our own configuration secrets for it?

jtkech replied:
@deanmarcussen 

Here @JoshLefebvre just override the `DefaultKeyVaultSecretManager` with a new `AzureKeyVaultSecretManager` and provide an `AddOrchardCoreAzureKeyVault()` extension that uses it and other config values.

But people can just follow the aspnet documentation and then use `new AzureKeyVaultSecretManager()` in place of `new DefaultKeyVaultSecretManager()`

So maybe just remove the `AddOrchardCoreAzureKeyVault()` that is just an example on how to use the new `AzureKeyVaultSecretManager` and let people follow the doc and do their own extension if they want.

xperiandri replied:
Do I understand your implementation right that you have added replacement of `---` to `_` along with default `--` to `:` behavior?

jtkech replied:
@xperiandri  Exactly, this is what @JoshLefebvre did ;)

sebastienros replied:
Is that something that is common? Either in OC or ASP.NET with Keyvault?

xperiandri commented:
Cool, so is only getting started guide left?

agriffard commented:
Please resolve conflict.

MarcBruins commented:
The preferred way of communicating with the Azure Keyvault for a web app is through Managed identity rather then a service principal with a object id and password. Can we make sure that that scenario is also supported? I can work on this if you like

Skrypt commented:
@MarcBruins I think that ultimately we should have an external OC Management tool that could be deployed as an Azure Function that would allow to configure all things related with the Azure Portal services. Here, what I don't like is the fact that OC configures it's own Azure Services. This can easily lead to security issues.

Also, I think that it could be a good way to start experiment things with the new Azure SDK.

agriffard commented:
Can you please fix the conflicts @JoshLefebvre?

agriffard commented:
Conflicts solved.

sebastienros commented on src/docs/reference/core/KeyVault.Azure/README.md:
````diff
@@ -0,0 +1,41 @@
+```json
+"OrchardCore_Azure_KeyVault": {
+    "KeyVaultName": "", // Set the name of your Azure Key Vault.
+    "AzureADApplicationId": "", // Set the Azure AD Application Id
````
Is that required when using managed identity in Azure? I believe we can also associated an Azure Web App for instance to be allowed to access a KV account without anything to setup.

mariojsnunes replied:
Yes, it's the AzureServiceTokenProvider. I got it working on my OrchardCore project.
Some notes:
I was only able to get in working in a .net standard project (which is fine).
I think there is a new API for .net core, but I couldn't get it to work at the time (6 months ago)

deanmarcussen commented:
Just to note this will need to move to the new azure library package https://docs.microsoft.com/en-us/dotnet/core/compatibility/3.1-5.0#azure-microsoft-prefixed-azure-integration-packages-removed

agriffard commented:
@JoshLefebvre Can you please apply the changes sugested by @deanmarcussen ?

JoshLefebvre commented:
@agriffard yes sorry for the hiatus. I have a lot more free time now and hope to actively start contributing to this project again. I will try to get this done today

JoshLefebvre commented:
@deanmarcussen  @agriffard so I've updated the Key Vault package to make use of the new Azure SDK package Azure.Extensions.AspNetCore.Configuration.Secrets. I have also added the [Azure.Identity](https://github.com/Azure/azure-sdk-for-net/blob/master/sdk/identity/Azure.Identity/README.md) package which is now the recommended way to authenticate with Azure resources via Active Directory token authentication and should resolve the issues people were mentioning above .

Azure Identity supports:
- Service principal authentication
- Managed identity authentication
- User principal authentication

agriffard commented:
Can you please review it @deanmarcussen ?

deanmarcussen commented:
Great, thanks for updating this @JoshLefebvre - they've definitely improved the libraries a lot since moving to the new Azure packages.

Couple of thoughts, and I will bring @jtkech in here as well, as he may agree / disagree

Because this is a program level configuration, I don't think it should be using the `OrchardCore` appsettings section. That would suggest it is configurable at a tenant level, but as it is wholly for the host, it isn't tenant configurable. (please correct me if I'm wrong)

I'm also thinking it's not a module. Because we can't load it modularly, for the same reasons.

Which makes it more like the Shells.Azure configuration, which we did as a seperate nuget package, as part of the `OrchardCore` folder, rather than as part of modules.

Lastly we need to pick a name. Because it's KeyVault Configuration - rather than KeyVault itself, which would have a read/write option, and we might be using it for read / write (i.e. it'd be good to do a `Shells.KeyVault` version at some point with read/write, and I'm hoping to do a Secrets KeyVault provider for the Secrets module https://github.com/OrchardCMS/OrchardCore/pull/7891, again for read/write). Both of those might connection to a different KeyVault - in the case of secrets, it might be a different vault for each tenant (as an option).

Azure have called it `Azure.Extensions.AspNetCore.Configuration.Secrets` - which is not helpful ;)

Would `OrchardCore.Configuration.KeyVault` be a good name?
or
`OrchardCore.Configuration.Azure` (I prefer KeyVault as it seems explicit and more obvious, but naming is hard)

jtkech commented:
@deanmarcussen okay i will take a look on it this night

jtkech commented:
@deanmarcussen 

Yes i agree, didn't see it was under the `OC.Modules` folder, need to be under the `OrchardCore` folder.

Then yes, good to have custom config / settings sources for all tenants and tenant specific, as we did from blob and database, but yes in a separate PR. So we already have `OrchardCore_Shells_Azure`, maybe rename it to `OrchardCore_Shells_Azure_Blob` so that we will be able to use `OrchardCore_Shells_Azure_KeyVault`. But that's okay to keep `OrchardCore_Shells_Azure` and then use `OrchardCore_Shells_KeyVault` ;)

Meanwhile for this PR, idem here i would suggest `OrchardCore.Configuration.Azure.KeyVault`, but maybe too long, in our solution there is no project with more than 3 terms, **so i also suggest `OrchardCore.Configuration.KeyVault`**.

agriffard commented:
Please merge dev and solve appsettings.json conflict.

deanmarcussen commented:
@JoshLefebvre sounds like @jtkech and I agree on the name, and to move it out of modules.

> I don't think it should be using the OrchardCore appsettings section

I withdraw my comment here, we already hosted the shells settings under `OrchardCore` so for consistency lets keep that the same. Even if it is only host only, not tenant based.

So just to move the folder, and rename thank you

jtkech commented:
I agree

JoshLefebvre commented:
Thanks @jtkech and @deanmarcussen. Yes agreed, makes much more sense in the OrchardCore module now that you explain it. I've also updated the name to OrchardCore.Configuration.KeyVault as suggested :)

agriffard commented:
Build retry

jtkech commented:
LGTM

agriffard commented:
Thank you. 163rd project in the solution.
